socketcan/libcanard/src/socketcan.c

```diff
@@ -162,43 +179,104 @@ int16_t socketcanPop(const SocketCANFD       fd,
-        struct canfd_frame cfd;
-        const ssize_t      read_size = read(fd, &cfd, sizeof(cfd));
-        if (read_size < 0)
+        // Initialize the message header used by recvmsg.
+        struct msghdr msg  = {0};           // Message header struct.
+        msg.msg_iov        = &iov;          // Scatter/gather array.
+        msg.msg_iovlen     = 1;             // Number of elements in the scatter/gather array.
+        msg.msg_control    = control;       // Ancillary data.
+        msg.msg_controllen = control_size;  // Ancillary data buffer length.
+
+        // Non-blocking receive messages from the socket and validate.
+        const ssize_t read_size = recvmsg(fd, &msg, MSG_DONTWAIT);
+        if (read_size <= 0)
         {
-            return getNegatedErrno();
+            const int neg_errno = getNegatedErrno();
+
+            // Return either the negated error code, or zero in case the socket is marked nonblocking and the receive
+            // operation would block. POSIX.1-2001 allows either EWOULDBLOCK or EAGAIN to be returned for this case,
+            // and does not require these constants to have the same value. Thus, we check for both possibilities.
+            free(control);
+            return (read_size < 0 && (neg_errno == -EWOULDBLOCK || neg_errno == -EAGAIN)) ? 0 : neg_errno;
         }
         if ((read_size != CAN_MTU) && (read_size != CANFD_MTU))
         {
+            free(control);
             return -EIO;
         }
-        if (cfd.len > payload_buffer_size)
+        if (sockcan_frame.len > payload_buffer_size)
         {
+            free(control);
             return -EFBIG;
         }
 
-        const bool valid = ((cfd.can_id & CAN_EFF_FLAG) != 0) &&  // Extended frame
-                           ((cfd.can_id & CAN_RTR_FLAG) == 0) &&  // Not RTR frame
-                           ((cfd.can_id & CAN_ERR_FLAG) == 0);    // Not error frame
+        const bool valid = ((sockcan_frame.can_id & CAN_EFF_FLAG) != 0) &&  // Extended frame
+                           ((sockcan_frame.can_id & CAN_ERR_FLAG) == 0) &&  // Not RTR frame
+                           ((sockcan_frame.can_id & CAN_RTR_FLAG) == 0);    // Not error frame
         if (!valid)
         {
+            free(control);
             return 0;  // Not an extended data frame -- drop silently and return early.
         }
 
+        const bool loopback_frame = (msg.msg_flags & (int) MSG_CONFIRM) != 0;  // NOLINT
+        if (loopback == NULL && loopback_frame)
+        {
+            free(control);
+            return 0;  // The loopback pointer is NULL and this is a loopback frame -- drop silently and return early.
+        }
+        if (loopback != NULL)
+        {
+            *loopback = loopback_frame;
+        }
+
+        // Obtain the CAN frame TAI time stamp from the kernel.
+        // Comparing to clock_gettime(CLOCK_TAI, &ts) proved that the time stamp is from the desired, monotonous
+        // TAI clock source.
```

Comment:
TAI is not monotonous -- it may jump and change rate to account for the local clock error. More importantly, the timestamps supplied by the kernel are, in fact, CLOCK_REALTIME, so you need to update the docs accordingly, replacing CLOCK_TAI with CLOCK_REALTIME. The kernel/POSIX documentation is a bit lacking here, unfortunately. https://stackoverflow.com/questions/60971556/which-clock-does-so-timestamp-socket-option-use

Reply:
Thanks for the clarification. I've read the post, followed some of the links therein, and update accordingly.
I'm still left with a couple questions regarding this (and an uneasy feeling):

- I just effectively changed the API behaviour from CLOCK_TAI to CLOCK_REALTIME in a way that won't be evident to someone just updating this file in their code base.

- In the UAVCAN spec, TAI is the preferred time system. Is there a way to convert REALTIME to TAI without silly overhead? Mainly, I have an instinctive unease about mixing time systems, and strive to keep things uniform (rule of least astonishment, and I am easily astonisch).

- This timestamping clock chosen in the kernel makes me wonder what will happen when a leap second occurs. Suddenly, CAN frames with a tight deadline might be considered expired when the nodes are still updating their clocks (because time sync is not instantaneous), yet the time providing node has already applied the leap second. (In my project, the time provider is the same as the main data processor). It seems to me that, right at this junction, all these tight deadline frames will be rejected until the node clocks have synced-up.

Reply:
> I just effectively changed the API behaviour from CLOCK_TAI to CLOCK_REALTIME in a way that won't be evident to someone just updating this file in their code base.

This is not quite a library, it's more of an example, but in general, you are right, we should add a changelog. Can you please add one in the header comment at `socketcan.h`?

> In the UAVCAN spec, TAI is the preferred time system. Is there a way to convert REALTIME to TAI without silly overhead? Mainly, I have an instinctive unease about mixing time systems, and strive to keep things uniform (rule of least astonishment, and I am easily astonished).

I am not sure. A long time ago I asked a very similar question: "how do I convert between CLOCK_MONOTONIC and CLOCK_REALTIME?" My research led me exactly nowhere and on Stackoverflow I only got stupid advice to query both and subtract one from the other, so I ended up writing a user-space estimator using the Olson time lag recovery algorithm. It was not fun. In this case it may be much easier because as long as the number of leap seconds does not change, the difference is constant, and it is always an integer number of seconds, so the silly method of querying both and subtracting might actually work (proper precautions against non-uniform read latency assumed).

Mixing time systems really does create confusion and you may shoot yourself in the foot easily but it doesn't make the code non-compliant so it's not a catastrophe.

> This timestamping clock chosen in the kernel makes me wonder what will happen when a leap second occurs. Suddenly, CAN frames with a tight deadline might be considered expired when the nodes are still updating their clocks (because time sync is not instantaneous), yet the time providing node has already applied the leap second. (In my project, the time provider is the same as the main data processor). It seems to me that, right at this junction, all these tight deadline frames will be rejected until the node clocks have synced-up.

Yes, this is why one should steer clear from CLOCK_REALTIME and UTC clock in general when real-time systems are concerned, but the Linux kernel gives us no choice. There are some proposals floating to abolish the leap second, by the way.

Reply:
I'll add the changelog.

This thus all seems like there is no reasonably ready solution. The query and subtract approach indeed doesn't pass the sniff test in this context. Doing it right is also very error prone. Seems that validation to the monotonic clock would be required to make sure that the sampled (UTC) times are right and no one just happened to have adjusted the clock an hour backward or any other such thing. Or there was no sudden re-sync to GPS on a smaller but still relevant time-scale that would be, essentially, hard/impossible to detect.

I recall interesting incidents, such as a military aircraft passing the international dateline the wrong way round and, well, you can guess what happened next... Once, I read part of a design document regarding C coding for such systems and there was a blanket ban on using anything to do with ctime.

A conclusion from all of this could be that could be to keep the time node clock in UTC (to avoid daylight savings) and just accept that every so often, during X ms while the time synchronization system updates these infrequent clock changes, data is lost. Just like what would happen during some EM interference event.

Not particularly happy with this, and I'll have to take it into account when I implement the time sync. Eg: that any change larger then a few ms should not just update the PLL, but result in an immediate re-set of the clock. (And purge of the TX/RX buffers or adjustment of the deadlines/timestamps in them by the clock change.) Or something like that. In any case, complications indeed.

Reply:
Maybe the actual root of the problem is that GNU/Linux is the wrong OS for real-time applications. Whether you timestamp in kernel space or in the user space (being able to use whatever time system you prefer), you get no guaranteed accuracy or latency bounds either way.